### PR TITLE
OAuth2 authentication provider for Nessie client

### PR DESCRIPTION
**Pull request by adutra**

### CODE REVIEW

File: api/client/src/main/java/org/projectnessie/client/http/impl/RequestContextImpl.java
Comment (snazy) [inline api/client/src/main/java/org/projectnessie/client/http/impl/RequestContextImpl.java:91]: Nit: `getOrElse("").equals(..."`

File: api/client/src/main/java/org/projectnessie/client/NessieConfigConstants.java
Comment (snazy) [inline api/client/src/main/java/org/projectnessie/client/NessieConfigConstants.java:49]: Can you add the `{@value #CONF_...}` javadoc content? Helps to have the config key handy in docs (if we ever have some automatic way to generate a page out of this one).

File: api/client/src/main/java/org/projectnessie/client/NessieConfigConstants.java
Comment (snazy) [inline api/client/src/main/java/org/projectnessie/client/NessieConfigConstants.java:65]: Shall we link to `https://en.wikipedia.org/wiki/ISO_8601#Durations` ?

File: api/client/src/main/java/org/projectnessie/client/NessieConfigConstants.java
Comment (snazy) [inline api/client/src/main/java/org/projectnessie/client/NessieConfigConstants.java:64]: *sigh* I suspected, that the standard(s) require that an expiration time is sent with the token - but it's only recommended :(

File: api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2Client.java
Comment (snazy) [inline api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2Client.java:49]: I suspect we should let `NessieAuthenticator` extend `AutoCloseable` w/ a default noop and close associated authenticators when the client's closed

File: api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2ClientParams.java
Comment (snazy) [inline api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2ClientParams.java:103]: Add `@Value.Check` and omit the call in `OAuth2Client`?

**Review by snazy — COMMENTED** [review]
Really nice PR!
Left a bunch of comment - got too curious ;)

File: api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2ClientParams.java
Comment (adutra) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2ClientParams.java:90]: >  if the user forces the use of the "old" URLConnection client.

That's already supported: if the user forces the old client, the old client will be used here as well.

> configured with additional options for TLS/SSL and/or timeouts

First there would be a chicken-and-egg problem here. But besides that, my fear is that the options won't necessarily be the same for this client and the main client:

1. Base uri and authentication are not the same;
2. Timeouts could be different;
3. SSL could have different cipher suites, etc.

I think we will need a whole bunch of new settings under `nessie.authentication.oauth2.*` if we want to fine-tune the OAuth client.

File: api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2Client.java
Comment (adutra) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2Client.java:49]: That was my first intention, but then I backed off because the changes are a bit invasive: you need to "bubble up" the call to `close()` up to all `NessieApi` impls (they are already `AutoCloseable`, but nothing was enforcing that). But if you are OK with that, let's do it.

File: api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2Client.java
Comment (adutra) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2Client.java:87]: I'm going down the "AutoCloseable" path, so I believe we won't need this hook anymore, which is way more elegant.

File: api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2ClientParams.java
Comment (adutra) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2ClientParams.java:104]: There is a few methods already in `HttpUtils` btw.

File: api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2Client.java
Comment (adutra) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2Client.java:49]: It must be visible by `OAuth2AuthenticationProvider` which is in the parent package. I will move the provider here then.

File: api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2Client.java
Comment (adutra) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2Client.java:125]: I don't have experience enough to tell. I will play a bit with keycloak and revisit this later.

File: api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2Client.java
Comment (snazy) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2Client.java]: Yea, `NessieApi` instances should always be properly closed from call sites. I'm (mostly) sure, that we always do that.

File: api/client/src/main/java/org/projectnessie/client/http/impl/jdk8/UrlConnectionClient.java
Comment (snazy) [inline api/client/src/main/java/org/projectnessie/client/http/impl/jdk8/UrlConnectionClient.java:58]: If it makes sense, feel free to make `HttpClient` an abstract base class

File: api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2Client.java
Comment (snazy) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2Client.java]: _sigh_ Just noting: I realized that Iceberg's `Catalog` itself does not extend `(Auto)Closeable`, but a bunch of `Catalog` implementations do implement `(Auto)Closeable` (e.g. `NessieCatalog` or `RESTCatalog`). I suspect that we have to bite the bullet and live with the fact that `Catalog`s and in turn OAuth clients do not get properly closed.

Maybe we need to implement something in a follow-up - just "brain dumping" (no need to take it seriously):
* Track the last use time of the OAuth token
* It some time elapsed (couple minutes?), stop refreshing and stop the thread pool
* It the OAuth client is used again, restart the thread pool if necessary

I suspect, we need some mechanism like that in the future, but it's probably out of scope of this PR.

File: api/client/src/test/resources/logback-test.xml
Comment (dimas-b) [inline api/client/src/test/resources/logback-test.xml:31]: nit: Do we still need debug logs in STDOUT?

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2AuthenticationProvider.java
Comment (dimas-b) [inline api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2AuthenticationProvider.java:62]: Would it be possible to add a test case for this to `AbstractOpenIdAuthentication` (if token type can be something other than `Bearer`)?

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2AuthenticationProvider.java
Comment (dimas-b) [inline api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2AuthenticationProvider.java:79]: I suppose it is safer to use `.toNNNCase(Locale.ROOT)`. We do not expect non-ASCII chars there, do we? At the same time some fancy locales can take chars out of the ASCII range on case conversion.

**Review by dimas-b — COMMENTED** [review]
Still not done with a full review... Posting some preliminary comments :)

File: api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2Client.java
Comment (adutra) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2Client.java]: That was my suspicion, and it's unfortunate indeed. In the meantime, I removed the shutdown hook completely. If that's OK let's indeed explore solutions in a follow-up PR as this one is already quite big.

File: api/client/src/main/java/org/projectnessie/client/http/impl/jdk8/UrlConnectionClient.java
Comment (adutra) [reply api/client/src/main/java/org/projectnessie/client/http/impl/jdk8/UrlConnectionClient.java:58]: For now it's just one method that needs implementation in each concrete impl: `close()`, so I think it's not (yet) justified to make this change.

File: api/client/src/test/resources/logback-test.xml
Comment (adutra) [reply api/client/src/test/resources/logback-test.xml:31]: Sorry that was a leftover.

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2AuthenticationProvider.java
Comment (adutra) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2AuthenticationProvider.java:62]: To be fair, I don't think it can be something else than `(B|b)earer`. I wonder if we shouldn't just remove this `capitalize` stuff and print `Bearer` all the time.

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2AuthenticationProvider.java
Comment (adutra) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2AuthenticationProvider.java:79]: It's only ASCII as per RFC 7230. Thus `Locale.ROOT` is fine.

File: api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2Client.java
Comment (snazy) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2Client.java]: Sure, let's do that independently of this PR. Mind opening an issue for it?

File: gradle/libs.versions.toml
Comment (snazy) [inline gradle/libs.versions.toml:217]: Nit: move up a bit (alphabetical order)

File: api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2Client.java
Comment (snazy) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2Client.java]: (Except that Iceberg `Catalog`s do not implement `AutoCloseable`, but some implementations do *sigh*)

File: api/client/src/main/java/org/projectnessie/client/http/impl/jdk8/UrlConnectionClient.java
Comment (snazy) [reply api/client/src/main/java/org/projectnessie/client/http/impl/jdk8/UrlConnectionClient.java:58]: Okay for me

File: api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java
Comment (adutra) [inline api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java:95]: FYI I'm not entirely satisfied with this test that is admittedly very long (9*15 seconds). Today I'm going to try to create other tests focusing on details of the refresh cycle and maybe reduce this one to just one run of 15 seconds. (I think it's important to at least have one test exercising the full refresh cycle.)

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2AuthenticationProvider.java
Comment (snazy) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2AuthenticationProvider.java:63]: I'd omit `capitalize()` completely.
For example, [RFC 6749 section 7.1](https://www.rfc-editor.org/rfc/rfc6749.html#section-7.1) mentions a `MAC` authentication scheme (work in process). Not saying we should think of that at this time.
But overall, it can't probably be anything else than `Bearer` - maybe we should add a runtime check like `tokenType.toLowerCase(ROOT).equals("bearer")`?

File: api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java
Comment (snazy) [reply api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java:95]: Yea - that's very long. Maybe one quick IT integration round plus a couple with a mocked HTTP backend?

File: api/client/src/main/java/org/projectnessie/client/http/HttpClientBuilder.java
Comment (snazy) [inline api/client/src/main/java/org/projectnessie/client/http/HttpClientBuilder.java:323]: O O! Do not move this here, the Json-View setting won't be applied.

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2AuthenticationProvider.java
Comment (adutra) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2AuthenticationProvider.java:63]: I like your suggestion, let's remove the capitalize stuff and replace with a check.

File: api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2ClientParams.java
Comment (snazy) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2ClientParams.java]: Could be. Let's defer to a follow-up, if necessary.

File: api/client/build.gradle.kts
Comment (adutra) [inline api/client/build.gradle.kts:84]: @snazy let me know if there is a better way to handle this.

File: api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2Client.java
Comment (adutra) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2Client.java]: There you go: https://github.com/projectnessie/nessie/issues/6586.

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java
Comment (snazy) [inline api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java:107]: Would it be a lot of effort to have this covered in a unit test? Does it make sense?
(I see, it's covered in the IT, so not really an issue.)

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java
Comment (snazy) [inline api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java:145]: Nit: Should `this.tokenRefreshFuture` be set to `null` in a `finally`?

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java
Comment (snazy) [inline api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java:157]: Hm - if the same password's used in multiple clients/authentictors  (probably never happens though, but tests maybe...), a `close()` would corrupt the password for other instances.

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/JwtToken.java
Comment (snazy) [inline api/client/src/main/java/org/projectnessie/client/auth/oauth2/JwtToken.java:94]: Would something like this make it overall simpler?

```java

  private <T> T payloadFieldValue(String field, Predicate<JsonNode> tester, Function<JsonNode, T> mapper) {
    JsonNode value = payload.get(field);
    if (value == null || !tester.test(value)) {
      return null;
    }
    return mapper.apply(value);
  }
```

and for this `getIssuer`:
```java
          k -> payloadFieldValue(ISS_CLAIM, JsonNode::isTextual, JsonNode::asText));
```

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2AuthenticationProvider.java
Comment (snazy) [inline api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2AuthenticationProvider.java:67]:
```suggestion
            "OAuth2 token type returned from the authenticating server must be 'Bearer', but was: " + token.getTokenType());
```

**Review by snazy — COMMENTED** [review]
(Only finished a first, incomplete round, will continue tomorrow)

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2ClientParams.java
Comment (adutra) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2ClientParams.java]: In fact I just changed that. My thinking is that this object is meant to be short-lived, so there is really no need to worry about passwords here.

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java
Comment (adutra) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java:157]: Good catch, but that doesn't happen anymore, because in `OAuth2ClientParams` we are using plain strings now. This class creates a new array when reading the password.

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java
Comment (snazy) [inline api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java:189]: Does it make sense to log the error here or any (runtime) exception in `refreshTokens`?
Guess, it can only be either the HTTP request failing or the response not being deserializable in `refreshTokens` (and `exchangeTokens`).

**Review by snazy — COMMENTED** [review]
Looks really great!

Just a few comments, nothing dramatic.

And great job on test coverage!

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2AuthenticationProvider.java
Comment (adutra) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2AuthenticationProvider.java:74]: Done in ce91cbd37 and 30fbbe1e4, although I had to move `TestNessieHttpClient.java` to the `org.projectnessie.client.http` package for visibility reasons. But it makes sense to have that test in that package imho since the class under test is also in that package.

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java
Comment (adutra) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java:107]: That should be possible with some mocking. Good idea.

Comment (adutra) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java:107]: Done in 91b5ca7c6.

File: api/client/src/main/java/org/projectnessie/client/http/HttpClientBuilder.java
Comment (adutra) [inline api/client/src/main/java/org/projectnessie/client/http/HttpClientBuilder.java:66]: I wonder: we should probably not add this filter by default. For the OAuth2 client, it doesn't make sense for instance (it's not harmful either, just conceptually wrong since the client is not talking to Nessie at all).

Comment (snazy) [reply api/client/src/main/java/org/projectnessie/client/http/HttpClientBuilder.java:66]: Yes, doesn't make sense for OAuth. Probably safer to not add it for OAuth and only for Nessie REST.

Comment (adutra) [reply api/client/src/main/java/org/projectnessie/client/http/HttpClientBuilder.java:66]: Actually scratch that, the OAuth client uses `HttpClient.Builder`, not `HttpClientBuilder`... naming is hard :-)

**Review by snazy — APPROVED** [review]
Great stuff!
Ship it (once CI's happy)!

**Pull request merged**
